File: src/tree.go
package tree

import (
	"fmt"
	"strings"
)

// Node представляет узел бинарного дерева
type Node struct {
	data  int
	left  *Node
	right *Node
}

// NewNode создает новый узел с указанным значением
func NewNode(value int) *Node {
	return &Node{
		data:  value,
		left:  nil,
		right: nil,
	}
}

// FullBinaryTree представляет полное бинарное дерево
type FullBinaryTree struct {
	root *Node
	size int
}

// NewFullBinaryTree создает новое пустое бинарное дерево
func NewFullBinaryTree() *FullBinaryTree {
	return &FullBinaryTree{
		root: nil,
		size: 0,
	}
}

// Insert добавляет значение в дерево
func (t *FullBinaryTree) Insert(value int) {
	t.root = t.insert(t.root, value)
	t.size++
}

// insert рекурсивно вставляет значение в дерево
func (t *FullBinaryTree) insert(node *Node, value int) *Node {
	if node == nil {
		return NewNode(value)
	}
	
	if value < node.data {
		node.left = t.insert(node.left, value)
	} else {
		node.right = t.insert(node.right, value)
	}
	
	return node
}

// Search ищет значение в дереве
func (t *FullBinaryTree) Search(value int) bool {
	return t.search(t.root, value)
}

// search рекурсивно ищет значение в дереве
func (t *FullBinaryTree) search(node *Node, value int) bool {
	if node == nil {
		return false
	}
	
	if node.data == value {
		return true
	}
	
	// Рекурсивный поиск в обоих поддеревьях
	return t.search(node.left, value) || t.search(node.right, value)
}

// IsFull проверяет, является ли дерево полным
func (t *FullBinaryTree) IsFull() bool {
	return t.isFull(t.root)
}

// isFull рекурсивно проверяет, является ли поддерево полным
func (t *FullBinaryTree) isFull(node *Node) bool {
	if node == nil {
		return true
	}
	
	// Если это лист
	if node.left == nil && node.right == nil {
		return true
	}
	
	// Если есть оба потомка, проверяем рекурсивно
	if node.left != nil && node.right != nil {
		return t.isFull(node.left) && t.isFull(node.right)
	}
	
	// Если только один потомок - дерево не полное
	return false
}

// Print выводит дерево в консоль
func (t *FullBinaryTree) Print() {
	t.printTree(t.root, 0)
	fmt.Println()
}

// printTree рекурсивно выводит дерево с отступами
func (t *FullBinaryTree) printTree(node *Node, depth int) {
	if node == nil {
		return
	}
	
	// Сначала правое поддерево
	t.printTree(node.right, depth+1)
	
	// Вывод текущего узла с отступами
	indent := strings.Repeat("    ", depth)
	fmt.Printf("%s%d\n", indent, node.data)
	
	// Затем левое поддерево
	t.printTree(node.left, depth+1)
}

// ToString возвращает строковое представление дерева (pre-order обход)
func (t *FullBinaryTree) ToString() string {
	var result strings.Builder
	t.toString(t.root, &result)
	return result.String()
}

// toString рекурсивно собирает строковое представление
func (t *FullBinaryTree) toString(node *Node, sb *strings.Builder) {
	if node == nil {
		return
	}
	
	sb.WriteString(fmt.Sprintf("%d ", node.data))
	t.toString(node.left, sb)
	t.toString(node.right, sb)
}

// Size возвращает количество элементов в дереве
func (t *FullBinaryTree) Size() int {
	return t.size
}

// Clear очищает дерево
func (t *FullBinaryTree) Clear() {
	t.clear(t.root)
	t.root = nil
	t.size = 0
}

// clear рекурсивно очищает память узлов
func (t *FullBinaryTree) clear(node *Node) {
	if node == nil {
		return
	}
	
	t.clear(node.left)
	t.clear(node.right)
	// В Go не нужно явно освобождать память
}

// Вспомогательные методы для обходов дерева

// InOrder возвращает значения при инфиксном обходе (левый-корень-правый)
func (t *FullBinaryTree) InOrder() []int {
	result := make([]int, 0, t.size)
	t.inOrder(t.root, &result)
	return result
}

func (t *FullBinaryTree) inOrder(node *Node, result *[]int) {
	if node == nil {
		return
	}
	
	t.inOrder(node.left, result)
	*result = append(*result, node.data)
	t.inOrder(node.right, result)
}

// PreOrder возвращает значения при префиксном обходе (корень-левый-правый)
func (t *FullBinaryTree) PreOrder() []int {
	result := make([]int, 0, t.size)
	t.preOrder(t.root, &result)
	return result
}

func (t *FullBinaryTree) preOrder(node *Node, result *[]int) {
	if node == nil {
		return
	}
	
	*result = append(*result, node.data)
	t.preOrder(node.left, result)
	t.preOrder(node.right, result)
}

// PostOrder возвращает значения при постфиксном обходе (левый-правый-корень)
func (t *FullBinaryTree) PostOrder() []int {
	result := make([]int, 0, t.size)
	t.postOrder(t.root, &result)
	return result
}

func (t *FullBinaryTree) postOrder(node *Node, result *[]int) {
	if node == nil {
		return
	}
	
	t.postOrder(node.left, result)
	t.postOrder(node.right, result)
	*result = append(*result, node.data)
}

// FindMin возвращает минимальное значение в дереве
func (t *FullBinaryTree) FindMin() (int, error) {
	if t.root == nil {
		return 0, fmt.Errorf("tree is empty")
	}
	
	node := t.root
	for node.left != nil {
		node = node.left
	}
	return node.data, nil
}

// FindMax возвращает максимальное значение в дереве
func (t *FullBinaryTree) FindMax() (int, error) {
	if t.root == nil {
		return 0, fmt.Errorf("tree is empty")
	}
	
	node := t.root
	for node.right != nil {
		node = node.right
	}
	return node.data, nil
}

File: src/tree.cpp
#include "../include/tree.h"

NodeT::NodeT(int value) {
    data = value;
    left = nullptr;
    right = nullptr;
}

FullBinaryTree::FullBinaryTree() {
    root = nullptr;
    size = 0;
}

FullBinaryTree::~FullBinaryTree() {
    clear(root);
}

void FullBinaryTree::print() {
    printTree(root, 0);
    cout << endl;
}

string FullBinaryTree::toString() {
    return _toString(root);
}

void FullBinaryTree::insert(int value) {
    root = _insert(root, value);
    size++;
}

NodeT* FullBinaryTree::_insert(NodeT* node, int value) {
    if (node == nullptr) {
        return new NodeT(value);
    }
    if (value < node->data) {
        node->left = _insert(node->left, value);
    } else if (value > node->data) {
        node->right = _insert(node->right, value);
    }
    // Если значение уже существует, ничего не делаем
    return node;
}

// УДАЛЕНИЕ УЗЛА
void FullBinaryTree::remove(int value) {
    root = _remove(root, value);
    if(root) size--;
}

NodeT* FullBinaryTree::_remove(NodeT* node, int value) {
    if (node == nullptr) return nullptr;
    
    if (value < node->data) {
        node->left = _remove(node->left, value);
    } else if (value > node->data) {
        node->right = _remove(node->right, value);
    } else {
        // Найден узел для удаления
        
        // Случай 1: узел без потомков или с одним потомком
        if (node->left == nullptr) {
            NodeT* temp = node->right;
            delete node;
            return temp;
        } else if (node->right == nullptr) {
            NodeT* temp = node->left;
            delete node;
            return temp;
        }
        
        // Случай 2: узел с двумя потомками
        // Находим минимальный элемент в правом поддереве
        NodeT* temp = _findMin(node->right);
        node->data = temp->data;
        node->right = _remove(node->right, temp->data);
    }
    return node;
}

// Поиск минимального элемента
NodeT* FullBinaryTree::_findMin(NodeT* node) {
    while (node && node->left != nullptr) {
        node = node->left;
    }
    return node;
}

// Улучшенный поиск (для BST)
bool FullBinaryTree::search(NodeT* node, int value) {
    if (node == nullptr) return false;
    if (node->data == value) return true;
    
    if (value < node->data) {
        return search(node->left, value);
    } else {
        return search(node->right, value);
    }
}

// НОВЫЙ МЕТОД: поиск с возвратом bool
bool FullBinaryTree::search(int value) {
    return search(root, value);
}

bool FullBinaryTree::isFull() {
    return isFull(root);
}

bool FullBinaryTree::isFull(NodeT* node) {
    if (node == nullptr) return true;
    
    if (node->left == nullptr && node->right == nullptr) return true;
    if (node->left != nullptr && node->right != nullptr) {
        return isFull(node->left) && isFull(node->right);
    }
    return false;
}

// НОВЫЙ МЕТОД: проверка на полноту (все узлы имеют 0 или 2 потомка)
bool FullBinaryTree::isPerfect() {
    int height = getHeight(root);
    return _isPerfect(root, height, 0);
}

int FullBinaryTree::getHeight(NodeT* node) {
    if (node == nullptr) return 0;
    return 1 + max(getHeight(node->left), getHeight(node->right));
}

bool FullBinaryTree::_isPerfect(NodeT* node, int height, int level) {
    if (node == nullptr) return true;
    
    if (node->left == nullptr && node->right == nullptr) {
        return (height == level + 1);
    }
    
    if (node->left == nullptr || node->right == nullptr) {
        return false;
    }
    
    return _isPerfect(node->left, height, level + 1) &&
           _isPerfect(node->right, height, level + 1);
}

string FullBinaryTree::_toString(NodeT* node) {
    if (node == nullptr) return "";
    ostringstream oss;
    oss << node->data << " ";
    oss << _toString(node->left);
    oss << _toString(node->right);
    return oss.str();
}

// НОВЫЙ МЕТОД: обход в порядке in-order
string FullBinaryTree::inOrder() {
    return _inOrder(root);
}

string FullBinaryTree::_inOrder(NodeT* node) {
    if (node == nullptr) return "";
    ostringstream oss;
    oss << _inOrder(node->left);
    oss << node->data << " ";
    oss << _inOrder(node->right);
    return oss.str();
}

void FullBinaryTree::printTree(NodeT* node, int depth) {
    if (node == nullptr) return;
    printTree(node->right, depth + 1);
    cout << setw(4 * depth) << " " << node->data << endl;
    printTree(node->left, depth + 1);
}

void FullBinaryTree::clear(NodeT* node) {
    if (node == nullptr) return;
    clear(node->left);
    clear(node->right);
    delete node;
}

// НОВЫЙ МЕТОД: очистка всего дерева
void FullBinaryTree::clear() {
    clear(root);
    root = nullptr;
    size = 0;
}

// НОВЫЙ МЕТОД: получение размера
size_t FullBinaryTree::getSize() const {
    return size;
}
